bookwyrm/views/list.py

```diff
@@ -100,6 +105,59 @@ def get(self, request, list_id):
 
         query = request.GET.get("q")
         suggestions = None
+
+        page_count = (
+            (book_list.listitem_set.filter(approved=True).all().count() - 1)
+            // PAGINATION
+        ) + 1
+        if page_count < 1:  # special case for List with no items
+            page_count = 1
+
+        # sort_by shall be "order" unless a valid alternative is given
+        sort_by = request.GET.get("sort_by", "order")
+        if sort_by not in ("order", "title", "rating"):
+            sort_by = "order"
+
+        # direction shall be "ascending" unless a valid alternative is given
+        direction = request.GET.get("direction", "ascending")
+        if direction not in ("ascending", "descending"):
+            direction = "ascending"
+
+        # page shall be an integer from 1 to the max page count
+        page = request.GET.get("page", "1")
+        try:
+            page = int(page)
+        except ValueError:
+            page = 1
+        if page < 1:
+            page = 1
+        elif page > page_count:
+            page = page_count
+
+        if sort_by == "order":
+            directional_sort_by = "order"
+            if direction == "descending":
+                directional_sort_by = "-" + directional_sort_by
+            items = book_list.listitem_set.filter(approved=True).order_by(
+                directional_sort_by
+            )
+        elif sort_by == "title":
+            directional_sort_by = "book__title"
+            if direction == "descending":
+                directional_sort_by = "-" + directional_sort_by
+            items = book_list.listitem_set.filter(approved=True).order_by(
+                directional_sort_by
+            )
+        elif sort_by == "rating":
+            directional_sort_by = "average_rating"
+            if direction == "descending":
+                directional_sort_by = "-" + directional_sort_by
+            items = book_list.listitem_set.annotate(
+                average_rating=Avg(Coalesce("book__review__rating", 3))
+            ).order_by(directional_sort_by)
+
+        paged_items = items[((page - 1) * PAGINATION) : (page * PAGINATION)]
```

Comment:
I was under the impression, possibly incorrect, that the `.all()` in a query renders the full list of results. Seen in the referred snippet:

```python
lists = (
            models.List.objects.annotate(
                item_count=Count("listitem", filter=Q(listitem__approved=True))
            )
            .filter(item_count__gt=0)
            .order_by("-updated_date")
            .distinct()
            .all()
        )
        ...
        paginated = Paginator(lists, 12)
```

So if the list has 1000 books, it will fetch all 1000 but then pass only 12 to the template via the Page object, I think. By slicing a non-rendered query I sought to be more efficient with the fetching.

Reply:
Oh, good catch -- This shouldn't use `.all()`! But the django paginator will handling slicing efficiently (as long as you don't do what I did here and undermine the entire benefit 😂 )

Reply:
Cool :smile:  I removed the accidental `.all()` on that line while I'm in here.

Reply:
much appreciated